d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java

```diff
@@ -531,6 +533,182 @@ public void done()
     }
   }
 
+  @Test
+  public void testGetClientWithOptimizedAffinityRouting() throws Exception
+  {
+    Map<String, LoadBalancerStrategyFactory<? extends LoadBalancerStrategy>> loadBalancerStrategyFactories =
+        new HashMap<>();
+    Map<String, TransportClientFactory> clientFactories = new HashMap<>();
+    List<String> prioritizedSchemes = new ArrayList<>();
+
+    MockStore<ServiceProperties> serviceRegistry = new MockStore<>();
+    MockStore<ClusterProperties> clusterRegistry = new MockStore<>();
+    MockStore<UriProperties> uriRegistry = new MockStore<>();
+
+    ScheduledExecutorService executorService = new SynchronousExecutorService();
+
+    loadBalancerStrategyFactories.put("degrader", new DegraderLoadBalancerStrategyFactoryV3());
+    clientFactories.put(PropertyKeys.HTTP_SCHEME, new DoNothingClientFactory());
+
+    SimpleLoadBalancerState state =
+        new SimpleLoadBalancerState(executorService,
+            uriRegistry,
+            clusterRegistry,
+            serviceRegistry,
+            clientFactories,
+            loadBalancerStrategyFactories);
+
+    SimpleLoadBalancer loadBalancer =
+        new SimpleLoadBalancer(state, 5, TimeUnit.SECONDS, _d2Executor);
+
+    FutureCallback<None> balancerCallback = new FutureCallback<>();
+    loadBalancer.start(balancerCallback);
+    balancerCallback.get();
+
+    URI uri = URI.create("http://test.qd.com:5678");
+    Map<Integer, PartitionData> partitionData = new HashMap<>(1);
+    partitionData.put(DEFAULT_PARTITION_ID, new PartitionData(1d));
+    Map<URI, Map<Integer, PartitionData>> uriData = new HashMap<>(2);
+    uriData.put(uri, partitionData);
+
+    prioritizedSchemes.add(PropertyKeys.HTTP_SCHEME);
+
+    Set<URI> bannedSet = new HashSet<>();
+    clusterRegistry.put("cluster-1", new ClusterProperties("cluster-1", Collections.emptyList(),
+        Collections.emptyMap(), bannedSet, NullPartitionProperties.getInstance()));
+
+    serviceRegistry.put("foo", new ServiceProperties("foo",
+        "cluster-1",
+        "/foo",
+        Arrays.asList("degrader"),
+        Collections.<String,Object>emptyMap(),
+        null,
+        null,
+        prioritizedSchemes,
+        null));
+    uriRegistry.put("cluster-1", new UriProperties("cluster-1", uriData));
+
+    URI expectedUri = URI.create("http://test.qd.com:5678/foo");
+    URIRequest uriRequest = new URIRequest("d2://foo/52");
+
+    RequestContext serviceContext = new RequestContext();
+    serviceContext.putLocalAttr(AffinityRoutingURIProvider.AFFINITY_ROUTING_URI_PROVIDER, new AffinityRoutingURIProvider() {
+      private final Map<String, URI> uriMap = new HashMap<>();
```

Comment:
seems the uri was not added to this map before  loadBalancer.getClient?

Reply:
The test is correct.. first call to getClient makes an entry into the map and we verify in the test